tempesta_fw/apm.c

```diff
@@ -0,0 +1,1041 @@
+		TFW_DBG3("%s: Percentile calculation incomplete.\n", __func__);
+		set_bit(TFW_APM_DATA_F_RECALC, &data->flags);
+	} else {
+		TFW_DBG3("%s: Percentile values may have changed.\n", __func__);
+		write_lock(&prcntl->rwlock);
+		memcpy(prcntl->pstats, pstats,
+		       prcntl->pstsz * sizeof(PrcntlStats));
+		atomic_inc(&data->stats.rdidx);
+		write_unlock(&prcntl->rwlock);
+	}
+}
+
+/*
+ * Calculate the latest percentiles if necessary.
+ * Runs periodically on timer. 
+ */
+static void
+tfw_apm_prcntl_fn(unsigned long fndata)
+{
+	TfwApmData *data = (TfwApmData *)fndata;
+
+	tfw_apm_calc(data);
+
+	smp_mb__before_atomic();
+	if (test_bit(TFW_APM_DATA_F_REARM, &data->flags))
+		mod_timer(&data->timer, jiffies + TFW_APM_TIMER_TIMEOUT);
+}
+
+/*
+ * Get the latest calculated percentiles.
+ *
+ * Return 0 if the percentile values didn't need recalculation.
+ * Return 1 if potentially new percentile values were calculated.
+ */
+int
+tfw_apm_stats(void *apmdata, Percentile *percentile)
+{
+	int rdidx, seq = percentile->seq;
+	TfwApmData *data = apmdata;
+	TfwApmPrcntl *prcntl;
+
+	BUG_ON(!apmdata);
+
+	smp_mb__before_atomic();
+	rdidx = (unsigned int)atomic_read(&data->stats.rdidx) % 2;
+	prcntl = &data->stats.prcntl[rdidx];
+
+	read_lock(&prcntl->rwlock);
+	memcpy(percentile->pstats, prcntl->pstats,
+	       percentile->pstsz * sizeof(PrcntlStats));
+	read_unlock(&prcntl->rwlock);
+	percentile->seq = rdidx;
+
+	return (seq != rdidx);
+}
+
+/*
+ * Verify that an APM Stats user using the same set of percentiles.
+ *
+ * Note: This module uses a single set of percentiles for all servers.
+ * All APM Stats users must use the same set of percentiles.
+ */
+int
+tfw_apm_percentile_verify(Percentile *percentile)
+{
+	int i;
+
+	if (percentile->pstsz != ARRAY_SIZE(tfw_apm_pstats))
+		return 1;
+	for (i = 0; i < percentile->pstsz; ++i)
+		if (percentile->pstats[i].ith != tfw_apm_pstats[i].ith)
+			return 1;
+	return 0;
+}
+
+static inline void
+__tfw_apm_update(TfwApmRBuf *rbuf, unsigned long jtstamp, unsigned long jrtime)
+{
+	int centry = (jtstamp / tfw_apm_jtmintrvl) % rbuf->rbufsz;
+	unsigned long jtmistart = jtstamp - (jtstamp % tfw_apm_jtmintrvl);
+	TfwApmRBEnt *crbent = &rbuf->rbent[centry];
+
+	tfw_apm_rbent_checkreset(crbent, jtmistart);
+	tfw_stats_update(&crbent->pcntrng, jiffies_to_msecs(jrtime));
+}
+
+void
+tfw_apm_update(void *apmdata, unsigned long jtstamp, unsigned long jrtime)
+{
+	BUG_ON(!apmdata);
+	__tfw_apm_update(&((TfwApmData *)apmdata)->rbuf, jtstamp, jrtime);
+}
+
+/*
+ * Destroy the specified APM ring buffer.
+ */
+void
+tfw_apm_destroy(void *apmdata)
+{
+	TfwApmData *data = apmdata;
+
+	clear_bit(TFW_APM_DATA_F_REARM, &data->flags);
+	smp_mb__after_atomic();
+	del_timer_sync(&data->timer);
+
+	kfree(data);
+}
+
+/*
+ * Initialize a ring buffer entry.
+ */
+static inline void
+tfw_apm_rbent_init(TfwApmRBEnt *rbent, unsigned long jtmistamp)
+{
+	memcpy(rbent->pcntrng.ctl, tfw_rngctl_init, sizeof(rbent->pcntrng.ctl));
```

Comment:
This function is only used when a ring buffer for a server is created and its entries are initialized. When everything is working, an entry is resent using `__tfw_apm_rbent_reset()` function.